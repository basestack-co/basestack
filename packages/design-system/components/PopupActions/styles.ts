import styled from "styled-components";
import { rem } from "polished";
import { position } from "styled-system";
import { scrollbar } from "../../styles";

export const Container = styled.div`
  ${position};
  background-color: ${({ theme }) => theme.popupActions.backgroundColor};
  box-shadow: ${({ theme }) => theme.shadow.elevation6};
  padding: ${({ theme }) => theme.spacing.s1};
  border-radius: 4px;
  width: ${rem("220px")};
  z-index: ${({ theme }) => theme.zIndex.popup};
`;

export const Col = styled.div`
  display: flex;
  flex-direction: column;
`;

export const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
  max-height: 264px;
  overflow-y: auto;
  ${scrollbar};
`;

export const Header = styled.div`
  display: flex;
  align-items: center;
  position: sticky;
  background-color: ${({ theme }) => theme.popupActions.backgroundColor};
  top: 0;
  padding: ${({ theme }) => theme.spacing.s1} ${({ theme }) => theme.spacing.s2}
    0 ${({ theme }) => theme.spacing.s2};
  will-change: transform;
  transform: translateZ(0);
`;

export const List = styled.ul`
  display: flex;
  flex-direction: column;
`;

export const ListItem = styled.li`
  display: flex;
  flex-direction: column;
`;

export const PopUpButton = styled.button`
  display: flex;
  align-items: center;
  background-color: ${({ theme }) => theme.popupActions.button.backgroundColor};
  border: none;
  padding: ${rem("8px")};
  cursor: pointer;
  border-radius: 4px;

  &:hover {
    background-color: ${({ theme }) =>
      theme.popupActions.button.hover.backgroundColor};
  }
  &:active {
    background-color: ${({ theme }) =>
      theme.popupActions.button.active.backgroundColor};
  }
`;
