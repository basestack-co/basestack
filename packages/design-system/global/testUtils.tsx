import { type RenderOptions, render } from "@testing-library/react";
import type React from "react";
// Styles
import { ThemeProvider } from "styled-components";
import theme from "../theme/lightTheme";

interface Props {
  children: React.ReactNode;
}

const WithThemeProvider = ({ children }: Props) => (
  <ThemeProvider theme={theme}>{children}</ThemeProvider>
);

const renderWithTheme = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, "wrapper">,
) => render(ui, { wrapper: WithThemeProvider, ...options });

export { renderWithTheme };
