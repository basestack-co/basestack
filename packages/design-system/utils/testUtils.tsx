import defaultTheme from "@basestack/design-system/theme/lightTheme";
import isPropValid from "@emotion/is-prop-valid";
import { RenderResult, render } from "@testing-library/react";
import React from "react";
import { StyleSheetManager, ThemeProvider } from "styled-components";

const Provider = ({
  children,
  theme = defaultTheme,
}: {
  children: React.ReactNode;
  theme: any;
}) => (
  <ThemeProvider theme={theme}>
    <StyleSheetManager shouldForwardProp={isPropValid}>
      {children}
    </StyleSheetManager>
  </ThemeProvider>
);

export const renderWithTheme = (
  ui: React.ReactElement,
  options?: any,
): RenderResult => render(ui, { wrapper: Provider, ...options });
