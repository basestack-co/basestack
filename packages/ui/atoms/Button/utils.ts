import { Size, Target, Type, Variant } from "@/components/atoms/Button/types";
import { css, DefaultTheme } from "styled-components";
import { darken, rem, transparentize } from "polished";

export const handleButtonSize = (size?: Size) => {
  switch (size) {
    case "large":
      return css`
        height: ${rem("60px")};
        font-size: ${rem("18px")};
        line-height: ${rem("25px")};
        padding: 0 ${rem("40px")};
      `;
    case "medium":
      return css`
        height: ${rem("54px")};
        font-size: ${rem("16px")};
        line-height: ${rem("22px")};
        padding: 0 ${rem("34px")};
      `;
    case "normal":
      return css`
        height: ${rem("48px")};
        font-size: ${rem("16px")};
        line-height: ${rem("22px")};
        padding: 0 ${rem("30px")};
      `;
    default:
    case "small":
      return css`
        height: ${rem("46px")};
        font-size: ${rem("14px")};
        line-height: ${rem("19px")};
        padding: 0 ${rem("24px")};
      `;
    case "xSmall":
      return css`
        height: ${rem("42px")};
        font-size: ${rem("14px")};
        line-height: ${rem("19px")};
        padding: 0 ${rem("20px")};
      `;
  }
};

export const handleButtonsColors = (theme: DefaultTheme) => ({
  primary: {
    outline: {
      color: theme.colors.primary,
      borderColor: theme.colors.primary,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.white,
      borderColor: "transparent",
      backgroundColor: theme.colors.primary,
    },
  },
  secondary: {
    outline: {
      color: theme.colors.secondary,
      borderColor: theme.colors.secondary,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.white,
      borderColor: "transparent",
      backgroundColor: theme.colors.secondary,
    },
  },
  tertiary: {
    outline: {
      color: theme.colors.gray3,
      borderColor: theme.colors.gray3,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.primary,
      borderColor: "transparent",
      backgroundColor: theme.colors.gray6,
    },
  },
  neutral: {
    outline: {
      color: theme.colors.gray2,
      borderColor: theme.colors.gray2,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.primary,
      borderColor: "transparent",
      backgroundColor: "transparent",
    },
  },
  success: {
    outline: {
      color: theme.colors.success,
      borderColor: theme.colors.success,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.white,
      borderColor: "transparent",
      backgroundColor: theme.colors.success,
    },
  },
  warning: {
    outline: {
      color: theme.colors.warning,
      borderColor: theme.colors.warning,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.white,
      borderColor: "transparent",
      backgroundColor: theme.colors.warning,
    },
  },
  danger: {
    outline: {
      color: theme.colors.error,
      borderColor: theme.colors.error,
      backgroundColor: "transparent",
    },
    filled: {
      color: theme.colors.white,
      borderColor: "transparent",
      backgroundColor: theme.colors.error,
    },
  },
});

export const darkenColors = (
  theme: DefaultTheme,
  amount: number,
  type: Type,
  variant: Variant,
  target: Target
) => {
  return darken(
    amount || 0.1,
    handleButtonsColors(theme)[type || "primary"][variant || "filled"][target]
  );
};

export const lightenColors = (
  theme: DefaultTheme,
  amount: number,
  type: Type,
  variant: Variant,
  target: Target
) => {
  return transparentize(
    amount || 0.6,
    handleButtonsColors(theme)[type || "primary"][variant || "filled"][target]
  );
};
