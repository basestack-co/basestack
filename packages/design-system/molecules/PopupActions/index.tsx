import { forwardRef, memo, useCallback } from "react";
import { PositionProps } from "styled-system";
import { Avatar, Button, ButtonVariant, Text } from "../../atoms";
import { Container, Header, List, ListItem, PopUpButton } from "./styles";
import { useTheme } from "styled-components";

export interface PopupItem {
  id: string;
  slug: string;
  onClick: () => void;
  text: string;
}

export interface PopupActionsProps extends PositionProps {
  /**
   * Popup title
   */
  title: string;
  /**
   * List of actions
   */
  items: Array<PopupItem>;
  /**
   * Button props
   */
  button: {
    text: string;
    onClick: () => void;
  };
  onCallback?: () => void;
  /**
   * Truncates list item text to one line
   */
  truncateText?: boolean;
}

const PopupActions = forwardRef<HTMLDivElement, PopupActionsProps>(
  (
    { items, button, title, onCallback, truncateText = true, ...props },
    ref,
  ) => {
    const theme = useTheme();

    const onHandleClick = useCallback(
      (onClick: () => void) => {
        onClick();

        if (typeof onCallback === "function") {
          onCallback();
        }
      },
      [onCallback],
    );

    return (
      <Container ref={ref} {...props}>
        <Header>
          <Text muted fontWeight={500}>
            {title}
          </Text>
        </Header>
        <List>
          <ListItem>
            {items &&
              items.map((item, index) => {
                return (
                  <PopUpButton
                    onClick={() => onHandleClick(item.onClick)}
                    key={`pop-up-button-${item.id}`}
                  >
                    <Avatar
                      round={false}
                      userName={item.text}
                      alt={`Select ${item.text} from the list`}
                      size="small"
                    />
                    <Text ml={theme.spacing.s2} lineTruncate={truncateText}>
                      {item.text}
                    </Text>
                  </PopUpButton>
                );
              })}
          </ListItem>
        </List>
        <Button
          mt={theme.spacing.s1}
          justifyContent="center"
          variant={ButtonVariant.Tertiary}
          fullWidth
          onClick={button.onClick}
        >
          {button.text}
        </Button>
      </Container>
    );
  },
);

PopupActions.displayName = "PopupActions";

export default memo(PopupActions);
