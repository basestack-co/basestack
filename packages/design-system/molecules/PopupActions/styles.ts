import styled from "styled-components";
import { rem } from "polished";
import { position } from "styled-system";
import { scrollbar } from "../../styles";

export const Container = styled.div`
  ${position};
  background-color: ${({ theme }) => theme.colors.white};
  box-shadow: ${({ theme }) => theme.shadow.elevation6};
  padding: ${({ theme }) => theme.spacing.s1};
  border-radius: 4px;
  width: ${rem("220px")};
  z-index: ${({ theme }) => theme.zIndex.popup};
`;

export const Header = styled.div`
  display: flex;
  align-items: center;
  height: 32px;
  padding: 0 ${({ theme }) => theme.spacing.s2};
`;

export const List = styled.ul`
  display: flex;
  flex-direction: column;
  max-height: 264px;
  overflow-y: scroll;
  ${scrollbar};
`;

export const ListItem = styled.li`
  display: flex;
  flex-direction: column;
`;

export const PopUpButton = styled.button`
  display: flex;
  align-items: center;
  background-color: transparent;
  border: none;
  padding: ${rem("8px")};
  cursor: pointer;
  border-radius: 4px;

  &:hover {
    background-color: ${({ theme }) => theme.colors.gray100};
  }
  &:active {
    background-color: ${({ theme }) => theme.colors.gray200};
  }
`;
