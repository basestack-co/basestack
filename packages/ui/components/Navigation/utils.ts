// Types
import { Translate } from "next-translate";
import { NextRouter } from "next/router";
// Utils
import { config } from "@basestack/utils";
import { signOut } from "next-auth/react";

export const getInternalLinks = (t: Translate, formId: string) => {
  return [
    {
      icon: "flag",
      activeText: "Submissions",
      to: `/form/${formId}/submissions`,
      text: t("internal.submissions"),
    },
    {
      icon: "flag",
      activeText: "Setup",
      to: `/form/${formId}/setup`,
      text: t("internal.setup"),
    },
    {
      icon: "settings",
      activeText: "Settings",
      to: `/form/${formId}/settings/general`,
      text: t("internal.settings"),
    },
  ];
};

export const getExternalLinks = (t: Translate) => {
  return [
    {
      icon: "description",
      activeText: "Documentation",
      to: config.urls.docs.flags.base,
      isExternal: true,
      text: t("external.docs"),
    },
    {
      icon: "link",
      activeText: "Github",
      to: config.urls.repo,
      isExternal: true,
      text: t("external.github"),
    },
  ];
};

export const getAvatarDropdownList = (
  t: Translate,
  router: NextRouter,
  onCreateForm: () => void,
) => {
  return [
    {
      icon: "add_circle",
      text: t("create.form"),
      onClick: onCreateForm,
      separator: true,
    },
    {
      icon: "settings",
      text: t("dropdown.settings"),
      onClick: () =>
        router.push({
          pathname: "/user/profile/settings",
        }),
    },
    {
      icon: "logout",
      text: t("dropdown.logout"),
      onClick: signOut,
    },
  ];
};
