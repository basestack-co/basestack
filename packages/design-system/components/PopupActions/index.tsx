import { forwardRef, memo, useCallback } from "react";
import { PositionProps } from "styled-system";
import Text from "../Text";
import Avatar from "../Avatar";
import { Button, ButtonVariant } from "../Button";
import { Container, Header, List, ListItem, PopUpButton } from "./styles";
import { useTheme } from "styled-components";

export interface PopupActionProps {
  id: string;
  slug: string;
  onClick: () => void;
  text: string;
}

export interface PopupActionsProps extends PositionProps {
  /**
   * Popup title
   */
  title: string;
  /**
   * List of actions
   */
  items: Array<PopupActionProps>;
  /**
   * Button props
   */
  button: {
    text: string;
    onClick: () => void;
  };
  onCallback?: () => void;
  /**
   * Truncates list item text to one line
   */
  truncateText?: boolean;
}

const PopupActions = forwardRef<HTMLDivElement, PopupActionsProps>(
  (
    { items, button, title, onCallback, truncateText = true, ...props },
    ref,
  ) => {
    const theme = useTheme();

    const onHandleClick = useCallback(
      (onClick: () => void) => {
        onClick();

        if (typeof onCallback === "function") {
          onCallback();
        }
      },
      [onCallback],
    );

    return (
      <Container ref={ref} {...props}>
        <Header>
          <Text muted fontWeight={500}>
            {title}
          </Text>
        </Header>
        {items && (
          <List>
            {items.map((item, index) => {
              return (
                <ListItem key={`pop-up-button-${item.id}`}>
                  <PopUpButton onClick={() => onHandleClick(item.onClick)}>
                    <Avatar
                      round={false}
                      userName={item.text}
                      alt={`Select ${item.text} from the list`}
                      size="xSmall"
                    />
                    <Text ml={theme.spacing.s2} lineTruncate={truncateText}>
                      {item.text}
                    </Text>
                  </PopUpButton>
                </ListItem>
              );
            })}
          </List>
        )}
        <Button
          mt={theme.spacing.s1}
          justifyContent="center"
          variant={ButtonVariant.Tertiary}
          fullWidth
          onClick={button.onClick}
        >
          {button.text}
        </Button>
      </Container>
    );
  },
);

PopupActions.displayName = "PopupActions";

export default memo(PopupActions);
