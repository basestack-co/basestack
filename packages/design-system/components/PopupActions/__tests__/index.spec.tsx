import { cleanup } from "@testing-library/react";
import { renderWithTheme } from "../../../utils/testUtils";
import PopupActions from "..";

describe("PopupActions tests", () => {
  afterEach(cleanup);

  test("should render PopupActions correctly", () => {
    const { asFragment } = renderWithTheme(
      <PopupActions
        position="absolute"
        top={0}
        left={0}
        title="Projects"
        items={[
          {
            id: "1",
            text: "Moon flags",
            slug: "moon-lags",
            onClick: jest.fn(),
          },
          {
            id: "2",
            text: "Teams kids",
            slug: "teams-kids",
            onClick: jest.fn(),
          },
        ]}
        button={{
          text: "Create Project",
          onClick: jest.fn(),
        }}
      />,
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
