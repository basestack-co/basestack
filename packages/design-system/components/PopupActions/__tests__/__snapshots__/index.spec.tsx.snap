// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PopupActions tests should render PopupActions correctly 1`] = `
<DocumentFragment>
  .c2 {
  height: 2.25rem;
  padding: 0 0.75rem;
  font-size: 0.875rem;
  position: relative;
  border: none;
  display: flex;
  align-items: center;
  font-weight: 500;
  cursor: pointer;
  border-radius: 4px;
  text-decoration: none;
  transition: background-color 0.1s ease-in-out,color 0.1s ease-in-out,border-color 0.1s ease-in-out;
  background-color: #EEEEEE;
  color: #000000;
  width: normal;
  height: normal;
  margin-top: 0.25rem;
  justify-content: center;
  width: 100%;
}

.c2:hover:not(:active, :disabled) {
  background-color: #E2E2E2;
}

.c2 .material-symbols-sharp,
.c2 .material-symbols-rounded {
  color: inherit;
}

.c2 svg,
.c2 svg g {
  fill: #000000;
}

.c2:disabled {
  cursor: default;
  opacity: 0.6;
}

.c0 {
  position: absolute;
  top: 0;
  left: 0;
  background-color: #FFFFFF;
  box-shadow: rgba(0, 0, 0, 0.25) 0px 25px 50px -12px;
  padding: 0.25rem;
  border-radius: 4px;
  width: 13.75rem;
  z-index: 109;
}

.c1 {
  display: flex;
  flex-direction: column;
  max-height: 264px;
  overflow-y: auto;
}

.c1::-webkit-scrollbar {
  width: 4px;
}

.c1::-webkit-scrollbar-track {
  background: #EEEEEE;
}

.c1::-webkit-scrollbar-thumb {
  background: #CBCBCB;
}

.c1::-webkit-scrollbar-thumb:hover {
  background: #AFAFAF;
}

<div
    class="c0"
    title="Projects"
  >
    <div
      class="c1"
    />
    <button
      class="c2"
      data-testid="button"
      mt="0.25rem"
      variant="tertiary"
    >
      <span
        class=""
      >
        Create Project
      </span>
    </button>
  </div>
</DocumentFragment>
`;
